File: src/kernel/arch/arm64/mmio.zig
//! Memory-mapped I/O helpers for ARM64.

pub inline fn read32(addr: usize) u32 {
    return @as(*volatile u32, @ptrFromInt(addr)).*;
}

pub inline fn write32(addr: usize, val: u32) void {
    @as(*volatile u32, @ptrFromInt(addr)).* = val;
}

pub inline fn write8(addr: usize, val: u8) void {
    @as(*volatile u8, @ptrFromInt(addr)).* = val;
}

File: rust/kernel/src/arch/aarch64/mmio.rs
use kernel::mmu::VirtualAddress;

// Volatile MMIO accessors. These are `unsafe` because the caller picks the
// address: a safe `VirtualAddress` can name anything, so soundness depends on
// the caller proving the address refers to a device register that tolerates the
// access width and side effects. Volatile prevents the compiler from eliding or
// reordering the access.

/// # Safety
/// `address` must refer to a 4-byte-aligned, currently mapped MMIO register
/// whose read has no side effects beyond the device's defined behavior.
pub unsafe fn read32(address: VirtualAddress) -> u32 {
    unsafe { core::ptr::read_volatile(address.get() as *const u32) }
}

/// # Safety
/// `address` must refer to a 4-byte-aligned, currently mapped MMIO register
/// that accepts a 32-bit write.
pub unsafe fn write32(address: VirtualAddress, value: u32) {
    unsafe { core::ptr::write_volatile(address.get() as *mut u32, value) };
}

/// # Safety
/// `address` must refer to a currently mapped MMIO register that accepts an
/// 8-bit write.
pub unsafe fn write8(address: VirtualAddress, value: u8) {
    unsafe { core::ptr::write_volatile(address.get() as *mut u8, value) };
}
